### ISSUE TITLE
Are there any plans to add typescript support to the repo?

### ISSUE THREAD

**Issue by morsh** [berniegp/mock-xmlhttprequest#13]
**Is this issue a *feature* request or a *bug* report?**
Feature request

**What is the current behavior?**
No autocomplete in typescript based projects

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

**What is the expected behavior?**
Autocomplete and type checking in development

**Environment in which you encounter the issue:**
Development

- mock-xmlhttprequest version:
- Node.js version: any
- OS version: any

**Did this work in a previous version of mock-xmlhttprequest (which)?**

**Do you intend to provide a pull request to address this issue?**

**Comment by berniegp** [comment]
I don't use TypeScript myself and have no plan to do so in the short term. I had a look at what adding TypeScript support implies and there seems to be 2 options:

1. Add a TypeScript definition file.
2. Convert this library in TypeScript.

Since this library is meant to be small and easily used in Node, I don't see option 2 happening. Option 1 could however be an unintrusive way to add type information. The JSDoc already covers all the function parameter types so it doesn't seem like a big job. Do you have any knowledge (or a pull request) to share regarding this?

Some references I found while researching this:

- https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html
- https://codeburst.io/https-chidume-nnamdi-com-npm-module-in-typescript-12b3b22f0724
- https://github.com/DefinitelyTyped/DefinitelyTyped
- https://www.mudbath.com.au/blogs/create-typescript-definitions-from-existing-javascript-libraries

**Comment by sheuertz** [comment]
I'll work on putting this into DefinitelyTyped for now, then will do a PR to pull it directly into this repo.

**Comment by berniegp** [comment]
Created the [typescript-decl](/berniegp/mock-xmlhttprequest/tree/typescript-decl) branch for this.

TODO:
- [x] getters and setters are missing on MockXhr (readyState, timeout, etc.)
- [x] create a function type for the url matchers in MockXhrServer
- [x] create a function type for the request handlers in MockXhrServer
- [x] create a return type for MockXhrServer.getRequestLog
- [x] is return type of newMockXhr() ok? It's a subclass of MockXhr
- [x] add tests in types/test

This list might not be exhaustive!

**Comment by tomasz-zablocki** [comment]
@berniegp

I probably won't manage maintaining another declarations package, but I can help with integrating dts-lint tooling into tests.

Answered your SO question btw

**Comment by berniegp** [comment]
@tomasz-zablocki I had a stab at integrating dst-lint but found that some rules seem really tailor-made for the DefinitelyTyped project (can't remember exactly, but something to do with directory structure if I recall correct;y). Maybe I missed a configuration option? In any case, I'd love to add dts-lint to this repo if it's possible. If you know of another repo that integrates it, I can have a look and replicate the setup.

I think that once I have correct declarations I'll be able to update them as new versions of this library come out. I'd be very grateful if you have time to provide hints for some or all of the TODO tasks I identified above.

**Comment by tomasz-zablocki** [comment]
@berniegp 

I assume the code is in **typescript-decl** branch.

Regarding dts-lint, you may actually be right regarding the folder structure. 
It can be replaced with [dts-jest](https://github.com/ikatyang/dts-jest).
I see that you're using mocha though and you probably don't want both testing frameworks at once.

I'll try to go through your TODO.
Please keep in mind that I only had a brief look at the library source - these are just hints that probably miss some important bits.

> * [ ]  getters and setters are missing on MockXhr (readyState, timeout, etc.)

Property accessors are not allowed in ambient context (declaration files). 
They are modeled by declaring class properties. 
Properties without setter accessor should be prefixed with `readonly` modifier.

> * [ ]  create a function type for the url matchers in MockXhrServer

Your matcher type is most likely: `type UrlMatcher = string | Regexp | (url: string)=>boolean`

> * [ ]  create a function type for the request handlers in MockXhrServer

Your handler type is likely:

```typescript
type Headers = Record<string, string|string[]>;

interface MockResponseHandler = {
   status: number;
   statusText: string;
   headers: Headers;
   body: string;
}

interface MockCallbackHandler {
    (xhr: MockXhr): void
} 

type Handler = Partial<MockResponseHandler> | MockCallbackHandler
```

> * [ ]  create a return type for MockXhrServer.getRequestLog

```typescript

// you can do this, which will introduce several types - one for each method and another one is union of all methods:
type HttpMethod = 'GET'|'POST'|'PUT'|'PATCH'|'HEAD'|'DELETE'|'CONNECT'|OPTIONS'|'TRACE' 
// if you want to allow arbitrary strings then you should replace it with:
type HttpMethod = string

interface RequestLogEntry {
      method: HttpMethod;
      url: string;
      headers: Headers;
      body?: string|null
    }}

class MockXhrServer {
    //...
    getRequestLog(): ReadonlyArray<RequestLogEntry>
}
```

> * [ ]  is return type of newMockXhr() ok? It's a subclass of MockXhr

Pretty sure it's ok, `LocalMockXhr` and `MockXhr` should have identical signatures.

> * [ ]  add tests in types/test

I don't know if there are any alternatives to **dts-lint/dts-jest**.  

Either of the two will be fine and become essential as soon as types become more complex.
Type correctness tests should probably have typical unit tests to complement them, especially when you're targeting different environments, for instance both tsc-compiled node services and web apps with babel pipelines. The difference in typescript output between `tsc` and `babel` may be minimal, but it's there.

If I'm correct, your library aims for API compliance with XmlHttpRequest specs. If that's true, then you should almost certainly make use of standard Typescript declarations like the ones for [DOM](https://github.com/microsoft/TypeScript/blob/master/lib/lib.dom.d.ts), probably saving yourself a lot of time.

**Comment by tomasz-zablocki** [comment]
Found [tsd](https://github.com/SamVerschueren/tsd) package today. 

It's not exactly the same as dts-lint or dts-jest, but might be your best shot, unless you're willing to swap mocha for jest.

It also seems to have streamlined and simple setup procedure.

**Comment by berniegp** [comment]
@tomasz-zablocki Thanks again for your help! Good find with the tsd package. It feels very "streamlined" and it seems to work well.

**Comment by berniegp** [comment]
@morsh can you test the branch typescript-decl in your project and let me know if it works well? I'll wait a couple of days before merging in master.

**Comment by berniegp** [comment]
Resolved in v5.0.0 (4fb1a19f180ee7f8a53247f34b17725505ed3495)

**Comment by morsh** [comment]
@berniegp tried to use the official documentation, and apparently the following fails:
(onSend) is not defined as part of type `MockXhr`
```
import * as MockXMLHttpRequest from 'mock-xmlhttprequest';
const MockXhr = MockXMLHttpRequest.newMockXhr();

MockXhr.onSend = (xhr) => {
  console.log('>>>>>>>>>>>>>>>>>>>>>>>>>>>>', xhr)
  const responseHeaders = { 'Content-Type': 'application/json' };
  const response = '{ "message": "Success!" }';
  xhr.respond(200, responseHeaders, response);
};

```

**Comment by berniegp** [comment]
@morsh Thanks for reporting this. This should be fixed in v5.1.0 now along with other TypeScript related errors.

**Comment by morsh** [comment]
Thanks!
